### PR TITLE
Improve DMA `pop` implementation

### LINKED ISSUE #1482: Bugs in DMA pop()
https://github.com/esp-rs/esp-hal/blob/822bef12c7df4dc00ca9b3f09ff60f15f0ae0198/esp-hal/src/dma/mod.rs#L740-L756

Found this while I was writing the camera driver, if `pop` is called with `data` of size less than `available`:
- You end up with a buffer overrun, right?
- All the remaining available data seems to be discarded too.

My proposed fix for this is.
```
fn pop(&mut self, data: &mut [u8]) -> Result<usize, DmaError> {
    let avail = self.available;

    if avail < data.len() {
        return Err(DmaError::Exhausted);
    }
    
    let dma_buf = unsafe {
        core::slice::from_raw_parts(self.read_buffer_start, avail)
    };
    data.copy_from_slice(&dma_buf[..data.len()]);

    self.available -= data.len();
    Ok(data.len())
}
```

What I don't understand is why `pop` returns a `usize` if it's always the same as `data`? Was pop meant to also work when the data buf is more than the available data?

### PR DESCRIPTION
**Pull request by bjoernQ**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
Fixes #1482 

#### Testing
We have a HIL test for I2S which is using this. It still works and even works when changing the code forcing it to get more than one descriptor's buffer

### CODE REVIEW

File: esp-hal/src/dma/mod.rs [hunk 1]
```diff
@@ -732,58 +730,65 @@ where
...
-                self.descriptors.as_mut_ptr()
-            } else {
-                next_descriptor
-            };
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
+                self.read_buffer_start = last_in_descr.buffer;
+                self.available = len;
+                return self.available;
+            }
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:750]: This is technically an existing bug but a bug none the less.

Isn't it incorrect here to give ownership of the descriptor (and buffer it points to) to the hardware before the application is done reading from it?
I think the `pop` method should do the work of `.set_owner(Owner::Dma);` since that's when the buffer is no longer read by the cpu/application.

File: esp-hal/src/dma/mod.rs [hunk 2]
```diff
@@ -732,58 +730,65 @@ where
...
+
+            if descr_ptr.is_null() {
+                break;
+            }
         }
 
-        self.available = 0;
+        self.last_seen_handled_descriptor_ptr = core::ptr::null_mut();
+        self.available = avail;
         Ok(data.len())
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:792]: Should this be `Ok(offset)`? Otherwise I don't really understand why this method returns a `usize`.

File: esp-hal/src/dma/mod.rs [hunk 3]
```diff
@@ -732,58 +730,65 @@ where
...
-        }
-
-        if self.available != 0 {
-            return self.available;
-        }
-
-        let descr_address = self.last_seen_handled_descriptor_ptr;
-        let mut dw0 = unsafe { descr_address.read_volatile() };
+        let last_in_dscr_ptr = self.rx_impl.last_in_dscr_address() as *mut DmaDescriptor;
+        let mut last_in_descr = unsafe { last_in_dscr_ptr.read_volatile() };
```
[the remark this answers is not in the document]
Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:734]: I think we really need volatile to keep the compiler from optimizing things out - the compiler doesn't know that something else (i.e. the DMA controller) modify it

File: esp-hal/src/dma/mod.rs (hunk 1 above)
Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:750]: Good idea - I agree

File: esp-hal/src/dma/mod.rs [hunk 4]
```diff
@@ -732,59 +730,66 @@ where
...
+            if self.last_seen_handled_descriptor_ptr != last_in_dscr_ptr {
+                self.available += len;
+                return self.available;
+            }
         }
 
         self.available
     }
 
     fn pop(&mut self, data: &mut [u8]) -> Result<usize, DmaError> {
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:759]: So just to confirm the behaviour of this method.

You're trying to read as many **whole** descriptors as you can into the provided `data` buffer?
Is the intent here to **not** support partial descriptor reads?

File: esp-hal/src/dma/mod.rs [hunk 5]
```diff
@@ -732,59 +730,66 @@ where
...
+        let mut descr_ptr = self.last_seen_handled_descriptor_ptr;
 
-        unsafe {
-            let dst = data.as_mut_ptr();
-            let src = self.read_buffer_start;
-            let count = self.available;
-            core::ptr::copy_nonoverlapping(src, dst, count);
+        while remaining != 0 && remaining >= avail {
+            unsafe {
+                let mut descr = descr_ptr.read_volatile();
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:768]: `descr_ptr` can be null if `available()` hasn't been called I think.

File: esp-hal/src/dma/mod.rs [hunk 6]
```diff
@@ -732,59 +730,66 @@ where
...
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
+                self.read_buffer_start = last_in_descr.buffer;
+                self.available = len;
+                return self.available;
+            }
+
+            if self.last_seen_handled_descriptor_ptr != last_in_dscr_ptr {
+                self.available += len;
+                return self.available;
+            }
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:753]: If the user calls `available()` multiple times before calling `pop()`, the length of the same descriptor will be incorrectly added multiple times right?

This method should probably check (and clear) the `IN_DONE` or `IN_SUC_EOF` interrupt like the `available()` implementation does on the Tx Channel.

File: esp-hal/src/dma/mod.rs (hunk 4 above)
Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:759]: Yes we only care about completely done descriptor buffers

File: esp-hal/src/dma/mod.rs (hunk 5 above)
Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:768]: true ... thanks

File: esp-hal/src/dma/mod.rs (hunk 6 above)
Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:753]: oh, that's embarrassing - I meant to use the `self.last_seen handled descriptor ptr != last_in_dscr_ptr` check for that but forgot to set `last_seen_handled_descriptor_ptr `

File: esp-hal/src/dma/mod.rs (hunk 4 above)
Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:759]: Supporting partial pop would be more complex because the descriptor buffers are not necessarily completely used - but I guess it's not much of a problem for users of the API. I added some documentation about these assumptions

Comment (Dominaezzz) [reply esp-hal/src/dma/mod.rs:752]: Ah no, I didn't mean reading a descriptor before the hardware is done writing to it. I meant reading part of of a fully written descriptor. At the moment descriptor sizes are about 4092, but a user maybe only want to pop 64 bytes at a time from the finished 4092 byte descriptor. (The tx channel has something like this I think)

Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:752]: For TX it's somewhat easier to do - "problem" really is that DMA doesn't always fully fill the descriptor's buffer. We could certainly support it but not sure if it will ever be a real issue for user code

Comment (Dominaezzz) [reply esp-hal/src/dma/mod.rs:752]: Yeah for most drivers it won't be important. For cameras though. partial reading will be important for JPEGs since the application won't know the size in advance.
I'm in favour of leaving this feature for a future PR though. No existing drivers need it right now.
I'll keep this in mind for my next comments.

File: esp-hal/src/dma/mod.rs [hunk 7]
```diff
@@ -732,59 +730,67 @@ where
...
-        if self.available != 0 {
-            return self.available;
+        let last_in_dscr_ptr = self.rx_impl.last_in_dscr_address() as *mut DmaDescriptor;
+        let last_in_descr = unsafe { last_in_dscr_ptr.read_volatile() };
+
+        if !last_in_descr.is_empty() && last_in_descr.owner() == Owner::Cpu {
+            let len = last_in_descr.len();
+
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:740]: You can skip this line I think, since you do this at the end of the function anyway

File: esp-hal/src/dma/mod.rs [hunk 8]
```diff
@@ -732,59 +730,67 @@ where
...
-        if dw0.owner() == Owner::Cpu && !dw0.is_empty() {
-            let descriptor_buffer = dw0.buffer;
-            let next_descriptor = dw0.next;
+    fn pop(&mut self, data: &mut [u8]) -> Result<usize, DmaError> {
+        let mut avail = self.available;
 
-            self.read_buffer_start = descriptor_buffer;
-            self.available = dw0.len();
+        let mut remaining = data.len();
+        let mut offset = 0;
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:756]: These two variables can be merged into one like so.
```suggestion
        let mut remaining = &mut data[..];
```

See next comment

File: esp-hal/src/dma/mod.rs [hunk 9]
```diff
@@ -732,59 +730,67 @@ where
...
-            dw0.set_suc_eof(false);
+        if descr_ptr.is_null() {
+            return Ok(0);
+        }
 
+        while remaining != 0 && remaining >= avail {
             unsafe {
-                descr_address.write_volatile(dw0);
+                let mut descr = descr_ptr.read_volatile();
+                let dst = data.as_mut_ptr().add(offset);
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:766]: Following from the previous comment, this can become like so.
```suggestion
                let dst = remaining.as_mut_ptr();
```

File: esp-hal/src/dma/mod.rs [hunk 10]
```diff
@@ -732,59 +730,67 @@ where
...
+                let count = descr.len();
+                core::ptr::copy_nonoverlapping(src, dst, count);
+
+                descr.set_owner(Owner::Dma);
+                descr.set_suc_eof(false);
+                descr.set_length(0);
+                descr_ptr.write_volatile(descr);
+
+                remaining -= count;
+                offset += count;
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:777]: Then this can become like so
```suggestion
                remaining = &mut remaining[count..];
```

File: esp-hal/src/dma/mod.rs [hunk 11]
```diff
@@ -732,59 +730,67 @@ where
...
+        let mut descr_ptr = self.last_seen_handled_descriptor_ptr;
 
-            dw0.set_owner(Owner::Dma);
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        if descr_ptr.is_null() {
+            return Ok(0);
+        }
 
+        while remaining != 0 && remaining >= avail {
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:763]: Then this can become like so
```suggestion
        while !remaining.is_empty() && remaining.len() >= avail {
```

Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:763]: The `remaining >= avail` is checking that the remaining space in the user provided buffer can fit all the available bytes. Which may include multiple descriptors right? (If `available()` is called multiple times before `pop()` is called).
I think this check is suppose to be `remaining >= descr.len()` as you only want to check that the user provided buffer can fit the next descriptor rather than all the available ones.

File: esp-hal/src/dma/mod.rs [hunk 12]
```diff
@@ -732,59 +730,67 @@ where
...
-            let src = self.read_buffer_start;
-            let count = self.available;
-            core::ptr::copy_nonoverlapping(src, dst, count);
+        if avail > 0 {
+            return Err(DmaError::BufferTooSmall);
         }
 
-        self.available = 0;
-        Ok(data.len())
+        self.last_seen_handled_descriptor_ptr = core::ptr::null_mut();
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:791]: Should this be?
```suggestion
        self.last_seen_handled_descriptor_ptr = descr_ptr;
```

Or maybe not that exactly but something about setting this to null (when `self.available` is not 0) seems wrong but I can't put my finger on it.

**Review by Dominaezzz — COMMENTED** [review]
Sorry about the endless stream of comments 😅, I'm just going through my mental notes of things I wanted to fix when I was writing the LCD_CAM driver.

**Comment by bjoernQ** [comment]
> Sorry about the endless stream of comments 😅, I'm just going through my mental notes of things I wanted to fix when I was writing the LCD_CAM driver.

No problem - you discovered actual bugs

File: esp-hal/src/dma/mod.rs [hunk 13]
```diff
@@ -732,59 +730,68 @@ where
...
+        if !last_in_descr.is_empty() && last_in_descr.owner() == Owner::Cpu {
+            let len = last_in_descr.len();
 
-            unsafe {
-                descr_address.write_volatile(dw0);
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.read_descriptor_ptr = last_in_dscr_ptr;
+                self.available = len;
+            } else if self.last_seen_handled_descriptor_ptr != last_in_dscr_ptr {
+                self.available += len;
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:743]: This code is assuming that `last_in_dscr_ptr` is equal to `self.last_seen_handled_descriptor_ptr.next`.
If the user takes a while to call `available()` again after calling it once, there may be multiple descriptors that now available.
So this here should be a while loop looking at all the descriptors between `self.last_seen_handled_descriptor_ptr` and `last_in_dscr_ptr`.

**Review by Dominaezzz — COMMENTED** [review]
My last comment

File: esp-hal/src/dma/mod.rs (hunk 13 above)
Comment (bjoernQ) [reply esp-hal/src/dma/mod.rs:743]: true - we already do something similar for TX 👍

File: esp-hal/src/dma/mod.rs [hunk 14]
```diff
@@ -733,58 +727,72 @@ where
...
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        let mut current_in_descr_ptr =
+            unsafe { self.last_seen_handled_descriptor_ptr.read_volatile() }.next;
+        let mut current_in_descr = unsafe { current_in_descr_ptr.read_volatile() };
 
-            unsafe {
-                descr_address.write_volatile(dw0);
-            }
+        while !current_in_descr.is_empty() && current_in_descr.owner() == Owner::Cpu {
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:740]: This check isn't needed. `pop` doesn't check it either.
```suggestion
        while current_in_descr.owner() == Owner::Cpu {
```

**Review by Dominaezzz — COMMENTED** [review]
LGTM

**Comment by bjoernQ** [comment]
I'm quite happy we now have a test which doesn't just run the overly optimistic happy-path but also tests some edge cases. Probably we can still improve on that but I think it's already in reasonably good shape

File: esp-hal/CHANGELOG.md [hunk 15]
```diff
@@ -24,6 +24,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Removed
 - uart: Removed `configure_pins` methods (#1592)
+- Removed `DmaError::Exhausted` error by improving the implementation of the `pop` function
```
Comment (MabezDev) [inline esp-hal/CHANGELOG.md:27]: Duplicated entry here

**Review by Dominaezzz — APPROVED** [review]
Ship it 🙂

**Review by jessebraham — APPROVED** [review]
I definitely still don't understand our DMA bits as well as I should, however these changes seem reasonable enough to me so happy to merge.

Perhaps @MabezDev could take a look at this too, just as a final check?

**Review by MabezDev — APPROVED** [review]
Looks good to me, thanks!
